chain/checkpointing/util.go

```diff
@@ -0,0 +1,285 @@
+
+	return TaggedHash("TapSighash", ss), nil
+}
+
+func PubkeyToTapprootAddress(pubkey []byte) string {
+	conv, err := bech32.ConvertBits(pubkey, 8, 5, true)
+	if err != nil {
+		fmt.Println("Error:", err)
+		log.Fatal("I dunno.")
+	}
+
+	// Add segwit version byte 1
+	conv = append([]byte{0x01}, conv...)
+
+	// regtest human-readable part is "bcrt" according to no documentation ever... (see https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki)
+	// Using EncodeM becasue we want bech32m... which has a new checksum
+	taprootAddress, err := bech32.EncodeM("bcrt", conv)
+	if err != nil {
+		fmt.Println(err)
+		log.Fatal("Couldn't produce our tapproot address.")
+	}
+	return taprootAddress
+}
+
+func PubkeyToTapprootAddressLegacy(pubkey []byte) string {
+	conv, err := bech32.ConvertBits(pubkey, 8, 5, true)
+	if err != nil {
+		fmt.Println("Error:", err)
+		log.Fatal("I dunno.")
+	}
+	// Add segwit version byte 1
+	conv = append([]byte{0x01}, conv...)
+
+	taprootAddressLegacy, err := bech32.Encode("bcrt", conv)
+	if err != nil {
+		fmt.Println(err)
+		log.Fatal("Couldn't produce our tapproot address.")
+	}
+	return taprootAddressLegacy
+}
+
+func ApplyTweakToPublicKeyTaproot(public []byte, tweak []byte) []byte {
+	group := curve.Secp256k1{}
+	s_tweak := group.NewScalar().SetNat(new(safenum.Nat).SetBytes(tweak))
+	p_tweak := s_tweak.ActOnBase()
+
+	P, _ := curve.Secp256k1{}.LiftX(public)
+
+	Y_tweak := P.Add(p_tweak)
+	YSecp := Y_tweak.(*curve.Secp256k1Point)
+	if !YSecp.HasEvenY() {
+		s_tweak.Negate()
+		p_tweak := s_tweak.ActOnBase()
+		Y_tweak = P.Negate().Add(p_tweak)
+		YSecp = Y_tweak.(*curve.Secp256k1Point)
+	}
+	PBytes := YSecp.XBytes()
+	return PBytes
+}
+
+func HashMerkleRoot(pubkey []byte, checkpoint []byte) []byte {
+	merkle_root := TaggedHash("TapLeaf", []byte{0xc0}, pubkey, checkpoint)
+	return merkle_root[:]
+}
+
+func HashTweakedValue(pubkey []byte, merkle_root []byte) []byte {
+	tweaked_value := TaggedHash("TapTweak", pubkey, merkle_root)
+	return tweaked_value[:]
+}
+
+func GenCheckpointPublicKeyTaproot(internal_pubkey []byte, checkpoint []byte) []byte {
+	merkle_root := HashMerkleRoot(internal_pubkey, checkpoint)
+	tweaked_value := HashTweakedValue(internal_pubkey, merkle_root)
+
+	tweaked_pubkey := ApplyTweakToPublicKeyTaproot(internal_pubkey, tweaked_value)
+	return tweaked_pubkey
+}
+
+func AddTaprootToWallet(taprootScript string) bool {
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"importaddress\", \"params\": [\"" + taprootScript + "\", \"\", true]}"
+	result := jsonRPC(payload)
+
+	if result["error"] == nil {
+		return true
+	}
+
+	err := result["error"].(map[string]interface{})
+	if err["code"].(float64) == -4 {
+		// Particular case where we are already in the process of adding the key
+		// because we are using 1 bitcoin node for all
+		return true
+	}
+
+	return false
+}
+
+func GetTaprootScript(pubkey []byte) string {
+	return "5120" + hex.EncodeToString(pubkey)
```

Comment:
The "5120" is not clear to me here

Reply:
It is the taproot script. The taproot script should start with 1 and is followed by the pubkey.